lib/Dialect/TritonGPU/Transforms/RemoveLayoutConversions.cpp

```diff
@@ -82,542 +82,676 @@ class DecomposeDotOperand : public mlir::RewritePattern {
+        if (def && forwardSlice.count(def) &&
+            (seen.insert(operand.get()).second == true))
+          queue.push_back(forOp.getRegionIterArg(operand.getOperandNumber()));
+      }
     }
   }
-  rewriter.setInsertionPoint(op);
-  if (op->getNumResults() == 0) {
-    Operation *newOp = cloneWithInferType(rewriter, op, mapping);
-    rewriter.eraseOp(op);
-    return;
-  }
-  auto *newOp = cloneWithInferType(rewriter, op, mapping);
-  auto newType = newOp->getResult(0).getType().cast<RankedTensorType>();
-  auto newCvtType = RankedTensorType::get(
-      newType.getShape(), newType.getElementType(), dstEncoding);
-  auto newCvt = rewriter.create<triton::gpu::ConvertLayoutOp>(
-      newOp->getLoc(), newCvtType, newOp->getResult(0));
-  rewriter.replaceOp(op, newCvt->getResults());
+  return false;
 }
 
-//
-class MoveConvertOutOfIf : public mlir::RewritePattern {
-public:
-  explicit MoveConvertOutOfIf(mlir::MLIRContext *context)
-      : mlir::RewritePattern(scf::IfOp::getOperationName(), 2, context) {}
+// Return true if the op is an op with a layout we don't want to change. We will
+// propagate the layout starting from anchor ops.
+static bool isLayoutAnchor(Operation *op) {
+  if (isa<triton::LoadOp, triton::StoreOp>(op))
+    return isExpensiveLoadOrStore(op);
+  if (isa<triton::DotOp, triton::AtomicRMWOp, triton::AtomicCASOp>(op))
+    return true;
+  return false;
+}
 
-  mlir::LogicalResult
-  matchAndRewrite(mlir::Operation *op,
-                  mlir::PatternRewriter &rewriter) const override {
-    auto ifOp = cast<scf::IfOp>(*op);
-    // If “scf.if” defines no values, “scf.yield” will be inserted implicitly.
-    // However, "scf.else" is not required to be present, so we need to check
-    // if it exists.
-    auto thenYield = ifOp.thenYield();
-    int numOps = thenYield.getNumOperands();
-    SmallVector<Value> newThenYieldOps = thenYield.getOperands();
-    SetVector<Operation *> thenCvts;
-    SmallVector<Type> newRetTypes;
-
-    bool hasElse = !ifOp.getElseRegion().empty();
-
-    scf::YieldOp elseYield;
-    SmallVector<Value> newElseYieldOps;
-    SetVector<Operation *> elseCvts;
-    if (hasElse) {
-      elseYield = ifOp.elseYield();
-      newElseYieldOps = elseYield.getOperands();
+void LayoutPropagation::initAnchorLayout() {
+  funcOp.walk([&](Operation *op) {
+    if (isLayoutAnchor(op)) {
+      for (auto result : op->getResults()) {
+        if (auto tensorType = result.getType().dyn_cast<RankedTensorType>()) {
+          // Workaround, don't popagate MMA layout unless there is a convert
+          // back to mma further down to avoid generating reduction with MMA
+          // layout that may have lower performance.
+          // This can be improved with more aggressive backward propagation.
+          if (tensorType.getEncoding().isa<triton::gpu::MmaEncodingAttr>() &&
+              !hasConvertToMMATransisitiveUse(op, tensorType.getEncoding()))
```

Comment:
Does it have to be called multiple times? `hasConvertToMMATransisitiveUse` has a `getForwardSlice` call that essentially may introduce duplicated computation.
Can you do a slice and store all the results that have been analyzed?

Reply:
This could be done by adding a cache but this is probably a premature optimization since in general the number of convert to mma layout would be small, there would be generally only 1 per dot and the overlap in the slice is likely to be small.